File: Service/LivestreamApp.Server/Streaming/Livestreams/Stream.cs
using LivestreamApp.Server.Shared.WebSockets;
using LivestreamApp.Server.Streaming.StreamingSources;
using Ninject.Extensions.Logging;
using System;

namespace LivestreamApp.Server.Streaming.Livestreams
{
    public class Stream
    {
        public string Id { get; set; }
        public string Title { get; set; }
        public string Description { get; set; }
        public string CountryCode { get; set; }
        public string Input { get; set; }
        public bool StartOnServiceStartup { get; set; }
        public bool IsStarted { get; private set; }
        public bool HasValidInputSource { get; set; }
        public bool IsInitialized { get; set; }

        private IStreamingSource Source { get; set; }

        private readonly IWebSocketServerAdapter _webSocketServerAdapter;
        private readonly IStreamingSourceFactory _streamingSourceFactory;
        private readonly ILogger _logger;

        public Stream(ILogger logger, IWebSocketServerAdapter webSocketServerAdapter,
            IStreamingSourceFactory streamingSourceFactory)
        {
            _logger = logger ?? throw new ArgumentNullException(nameof(logger));
            _streamingSourceFactory =
                streamingSourceFactory ?? throw new ArgumentNullException(nameof(streamingSourceFactory));
            _webSocketServerAdapter =
                webSocketServerAdapter ?? throw new ArgumentNullException(nameof(webSocketServerAdapter));
        }

        public void Initialize()
        {
            Source = _streamingSourceFactory.GetStreamingSourceByDeviceId(Input);
            HasValidInputSource = Source.HasValidDevice();
            if (!HasValidInputSource) _logger.Warn($"Stream {Id} has invalid input source.");
            IsInitialized = true;
        }

        public void Start()
        {
            if (IsInitialized && HasValidInputSource && StartOnServiceStartup)
            {
                Source.StartStreaming();
                _webSocketServerAdapter.AddStreamingWebSocketService($"/streams/{Id}", Source);
                _webSocketServerAdapter.AddLoggingWebSocketService($"/streams/log/{Id}", Source);
                IsStarted = true;
            }
        }

        public void Stop()
        {
            if (IsStarted)
            {
                Source.StopStreaming();
                _webSocketServerAdapter.RemoveWebSocketService($"/streams/{Id}");
                _webSocketServerAdapter.RemoveWebSocketService($"/streams/log/{Id}");
                IsStarted = false;
            }
        }
    }
}

File: Service/src/lib/streams/stream.ts
import { StreamData } from "./stream-data";
import { Logger } from "../util/logger";
import { StreamingSource } from "./streaming-source";
import { inject } from "inversify";

/**
 * Class representing a live stream
 */
export class Stream {

    public get id(): string {
        return this._streamData.id;
    }

    public get data(): StreamData {
        return this._streamData;
    }

    private _isStarted: boolean;

    private _source: StreamingSource;

    public get hasValidSource(): boolean {
        return this._source.hasValidDevice;
    }

    constructor(@inject("Logger") private _logger: Logger,
        @inject("StreamingSourceFactory") streamingSourceFactory: (deviceId: string, stream: Stream) => StreamingSource,
        private _streamData: StreamData) {
        this._source = streamingSourceFactory(_streamData.deviceId, this);

        if (!this._source.hasValidDevice) {
            this._logger.warn(`Stream ${this.id} has invalid device and will not be startable.`);
        }

        _logger.debug(`Loaded stream ${JSON.stringify(_streamData)}.`);
    }

    public start(): void {
        if (!this._isStarted && this.hasValidSource) {
            this._source.startStreaming();
            this._logger.info(`Started stream ${this._streamData.id}.`);
            this._isStarted = true;
        }
    }

    public stop(): void {
        if (this._isStarted) {
            this._source.stopStreaming();
            this._logger.info(`Stopped stream ${this._streamData.id}.`);
            this._isStarted = false;
        }
    }
}
